core/src/main/java/org/elasticsearch/common/util/BigArrays.java

```diff
@@ -464,6 +465,18 @@ public CircuitBreakerService breakerService() {
     }
 
     /**
+     * Allocate a new {@link BytesPage}.
+     */
+    public BytesPage newBytePage() {
```

Comment:
Now I am starting to wrap my head around this and I think I have an idea of how we can simplify this more. We are currently using this concrete `BytesPage` instead of `org.elasticsearch.common.util.ByteArray` which is due to the lack of accessing the underlying bytes. Yet, using `ByteArray` might have some beauty to it, especially since it allows to have a single instance that we can simply resize by calling `BigArrays#resize` instead of doing all the accounting ourself. So in-order to do that I think we would need to add a couple of methods to `ByteArray` for instance this:

```Java
public interface ByteArray extends BigArray {

 //... all the existing methods
 
 // this sets the page for the given offset to the bytes ref without materializing it
 // we can then in-turn wrap this with a ByteBuffer to read from NIO   
 public void getPageReference(int offset, BytesRef ref);

  // releases all allocated byte pages until the given offset
  // the bytes at the given offset are not accessible at offset `0`
 public void shrinkTo(int offset);

 // this returns a bytes reference to the given offset with the given length without releasing etc.
 public BytesReference toBytesReference(int offset, int length);
}
```

this way I think we can keep all the logic of releaseing and advancing to the `ByteArray` utils and we are basically just moving along with the messages. I think this would make a lot of code much simpler in here. WDYT?

Reply:
I'm not really opposed to modifying `ByteArray` to fit our needs. I even considered using it when I first worked on this PR but I did not want to do much modification to it at the time, so I held off.

I think in regards to:

`public BytesReference toBytesReference(int offset, int length);`

Both of those pieces of functionality should be added pretty trivially using `PagedBytesReference`.

I don't really think `public void getPageReference(int offset, BytesRef ref)` is needed as `PagedBytesReference` returns a `BytesRef` iterator that will not materialize bytes. And most nio apis support vectorized operations.

`shrinkTo` is more complicated as it does not look like the `ByteArray` was designed to be shrunk. Nor is it a ring buffer which means that every time you slice a message off the head (and then shrink) you have to rearrange the stuff in the arrays.